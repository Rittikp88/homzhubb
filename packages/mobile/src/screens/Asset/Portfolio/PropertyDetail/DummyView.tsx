import React from 'react';
import { View, StyleSheet } from 'react-native';
import { Text } from '@homzhub/common/src/components';
import { theme } from '@homzhub/common/src/styles/theme';

// TODO: Delete this file once all tab screens are ready
const DummyView = (): React.ReactElement => {
  return (
    <View style={styles.container}>
      <Text type="large">Coming Soon...</Text>
    </View>
  );
};

export default DummyView;

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    paddingVertical: 50,
    backgroundColor: theme.colors.white,
  },
});
