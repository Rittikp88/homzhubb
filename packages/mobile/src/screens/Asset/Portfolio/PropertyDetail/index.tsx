import React, { PureComponent } from 'react';
import { ScrollView, StyleSheet, TouchableOpacity, View } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { isEmpty } from 'lodash';
import { AlertHelper } from '@homzhub/common/src/utils/AlertHelper';
import { DateUtils } from '@homzhub/common/src/utils/DateUtils';
import { ErrorUtils } from '@homzhub/common/src//utils/ErrorUtils';
import { FunctionUtils } from '@homzhub/common/src/utils/FunctionUtils';
import { PortfolioNavigatorParamList } from '@homzhub/mobile/src/navigation/PortfolioStack';
import { AssetRepository } from '@homzhub/common/src/domain/repositories/AssetRepository';
import { PortfolioRepository } from '@homzhub/common/src/domain/repositories/PortfolioRepository';
import { AssetActions } from '@homzhub/common/src/modules/asset/actions';
import { CommonActions } from '@homzhub/common/src/modules/common/actions';
import { RecordAssetActions } from '@homzhub/common/src/modules/recordAsset/actions';
import { SearchActions } from '@homzhub/common/src/modules/search/actions';
import { OfferActions } from '@homzhub/common/src/modules/offers/actions';
import { AssetSelectors } from '@homzhub/common/src/modules/asset/selectors';
import { PortfolioSelectors } from '@homzhub/common/src/modules/portfolio/selectors';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { theme } from '@homzhub/common/src/styles/theme';
import { Text } from '@homzhub/common/src/components/atoms/Text';
import { BottomSheet } from '@homzhub/common/src/components/molecules/BottomSheet';
import { FullScreenAssetDetailsCarousel } from '@homzhub/mobile/src/components';
import Menu, { IMenu } from '@homzhub/mobile/src/components/molecules/Menu';
import PropertyConfirmationView from '@homzhub/mobile/src/components/molecules/PropertyConfirmationView';
import AssetCard from '@homzhub/mobile/src/components/organisms/AssetCard';
import NotificationTab from '@homzhub/mobile/src/screens/Asset/Portfolio/PropertyDetail/NotificationTab';
import { UserScreen } from '@homzhub/mobile/src/components/HOC/UserScreen';
import { Asset } from '@homzhub/common/src/domain/models/Asset';
import { Attachment } from '@homzhub/common/src/domain/models/Attachment';
import { ISetAssetPayload } from '@homzhub/common/src/modules/portfolio/interfaces';
import {
  ClosureReasonType,
  IClosureReasonPayload,
  IListingParam,
  IPropertyDetailPayload,
  ListingType,
} from '@homzhub/common/src/domain/repositories/interfaces';
import { NavigationScreenProps, ScreensKeys, UpdatePropertyFormTypes } from '@homzhub/mobile/src/navigation/interfaces';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { Routes, Tabs, IRoutes } from '@homzhub/common/src/constants/Tabs';
import { ISelectedAssetPlan } from '@homzhub/common/src/domain/models/AssetPlan';
import { Filters } from '@homzhub/common/src/domain/models/AssetFilter';
import { IFilter } from '@homzhub/common/src/domain/models/Search';
import { IChatPayload } from '@homzhub/common/src/modules/common/interfaces';
import { ICurrentOffer, IOfferCompare } from '@homzhub/common/src/modules/offers/interfaces';
import { IGetAssetPayload } from '@homzhub/common/src/modules/asset/interfaces';
import { EmptyState } from '@homzhub/common/src/components/atoms/EmptyState';

enum MenuItems {
  EDIT_LISTING = 'EDIT_LISTING',
  EDIT_PROPERTY = 'EDIT_PROPERTY',
  DELETE_PROPERTY = 'DELETE_PROPERTY',
  MANAGE_TENANT = 'MANAGE_TENANT',
  EDIT_LEASE = 'EDIT_LEASE',
}

interface IStateProps {
  assetPayload: ISetAssetPayload;
  asset: Asset | null;
}

interface IDispatchProps {
  setAssetId: (payload: number) => void;
  setSelectedPlan: (payload: ISelectedAssetPlan) => void;
  getAssetById: () => void;
  clearAsset: () => void;
  clearChatDetail: () => void;
  clearMessages: () => void;
  clearState: () => void;
  setEditPropertyFlow: (payload: boolean) => void;
  toggleEditPropertyFlowBottomSheet: (payload: boolean) => void;
  setCurrentChatDetail: (payload: IChatPayload) => void;
  setCurrentOfferPayload: (payload: ICurrentOffer) => void;
  getAsset: (payload: IGetAssetPayload) => void;
  setFilter: (payload: IFilter) => void;
  setCompareDetail: (payload: IOfferCompare) => void;
}

interface IDetailState {
  propertyData: Asset;
  attachments: Attachment[];
  isFullScreen: boolean;
  activeSlide: number;
  isLoading: boolean;
  isDeleteProperty: boolean;
  isFromTenancies: boolean | null;
}

type libraryProps = NavigationScreenProps<PortfolioNavigatorParamList, ScreensKeys.PropertyDetailScreen>;
type Props = WithTranslation & libraryProps & IStateProps & IDispatchProps;

export class PropertyDetailScreen extends PureComponent<Props, IDetailState> {
  public focusListener: any;

  constructor(props: Props) {
    super(props);
    const {
      route: { params },
    } = props;
    this.state = {
      propertyData: {} as Asset,
      isFullScreen: false,
      activeSlide: 0,
      attachments: [],
      isLoading: false,
      isDeleteProperty: false,
      isFromTenancies: params?.isFromTenancies ?? null,
    };
  }

  public componentDidMount = (): void => {
    const { navigation } = this.props;

    this.focusListener = navigation.addListener('focus', () => {
      this.getAssetDetail().then();
    });
  };

  public componentDidUpdate = (prevProps: Readonly<Props>): void => {
    const { asset, setCompareDetail } = this.props;
    if (JSON.stringify(asset) !== JSON.stringify(prevProps.asset)) {
      if (asset && asset.leaseTerm) {
        setCompareDetail({
          rent: asset.leaseTerm.expectedPrice,
          deposit: asset.leaseTerm.securityDeposit,
          incrementPercentage: asset.leaseTerm.annualRentIncrementPercentage ?? 0,
        });
      }

      if (asset && asset.saleTerm) {
        setCompareDetail({
          price: Number(asset.saleTerm.expectedPrice) ?? 0,
          bookingAmount: Number(asset.saleTerm.expectedBookingAmount) ?? 0,
        });
      }
    }
  };

  public componentWillUnmount(): void {
    const { clearAsset } = this.props;
    this.focusListener();
    clearAsset();
  }

  public render = (): React.ReactNode => {
    const {
      t,
      route: { params },
    } = this.props;
    const { propertyData, isLoading, isDeleteProperty, isFromTenancies } = this.state;

    const { assetStatusInfo, id } = propertyData;
    const isOccupied = assetStatusInfo?.tag.label === Filters.OCCUPIED;
    const menuItems = this.getMenuList(assetStatusInfo?.isListingPresent ?? false, isOccupied);
    const onPressAction = (payload: IClosureReasonPayload, param?: IListingParam): void =>
      this.handleAction(propertyData, payload, param);

    const title = params && params.isFromDashboard ? t('assetDashboard:dashboard') : t('portfolio');
    const isMenuIconVisible = assetStatusInfo?.tag.label !== Filters.EXPIRING && menuItems.length > 0;

    return (
      <View style={styles.flexOne}>
        <UserScreen
          title={title}
          scrollEnabled
          pageTitle={t('propertyDetails')}
          backgroundColor={theme.colors.background}
          headerStyle={styles.background}
          loading={isLoading}
          onBackPress={this.handleIconPress}
          rightNode={isMenuIconVisible ? this.renderRightNode(menuItems) : undefined}
          keyboardShouldPersistTaps
        >
          {!isEmpty(propertyData) ? (
            <>
              <AssetCard
                assetData={propertyData}
                isDetailView
                isFromTenancies={isFromTenancies ?? undefined}
                enterFullScreen={this.onFullScreenToggle}
                onCompleteDetails={this.onCompleteDetails}
                onOfferVisitPress={FunctionUtils.noop}
                onHandleAction={onPressAction}
                containerStyle={styles.card}
              />
              {this.renderTabView()}
              {!isFromTenancies && <NotificationTab propertyId={id} assetStatusInfo={assetStatusInfo} />}
            </>
          ) : (
            <EmptyState title={t('common:noDataAvailable')} icon={icons.portfolio} />
          )}
        </UserScreen>
        {this.renderFullscreenCarousel()}
        <BottomSheet
          visible={isDeleteProperty}
          headerTitle={t('property:deleteProperty')}
          sheetHeight={460}
          onCloseSheet={this.onCloseDeleteView}
        >
          <PropertyConfirmationView
            propertyData={propertyData}
            description={t('deletePropertyDescription')}
            message={t('deleteConfirmation')}
            onCancel={this.onCloseDeleteView}
            onContinue={(): Promise<void> => this.onDeleteProperty(propertyData.id)}
          />
        </BottomSheet>
      </View>
    );
  };

  private renderRightNode = (menuItems: IMenu[]): React.ReactElement => {
    const { t } = this.props;
    return <Menu data={menuItems} onSelect={this.onSelectMenuItem} optionTitle={t('property:propertyOption')} />;
  };

  private renderTabView = (): React.ReactElement | null => {
    const { isFromTenancies } = this.state;

    if (isFromTenancies) {
      return null;
    }

    return (
      <ScrollView horizontal showsHorizontalScrollIndicator={false} style={styles.tabView}>
        {Routes.map((route: IRoutes, index) => {
          return (
            <TouchableOpacity key={index} style={styles.tabItem} onPress={(): void => this.handleTabAction(route.key)}>
              <Icon name={route.icon || ''} color={theme.colors.darkTint3} size={26} />
              <Text type="small" style={styles.label}>
                {route.title}
              </Text>
            </TouchableOpacity>
          );
        })}
      </ScrollView>
    );
  };

  private renderFullscreenCarousel = (): React.ReactNode => {
    const { isFullScreen, activeSlide, attachments } = this.state;
    if (!isFullScreen) return null;
    return (
      <FullScreenAssetDetailsCarousel
        onFullScreenToggle={this.onFullScreenToggle}
        activeSlide={activeSlide}
        data={attachments}
        updateSlide={this.updateSlide}
      />
    );
  };

  private onFullScreenToggle = (attachments?: Attachment[]): void => {
    const { isFullScreen } = this.state;
    this.setState({ isFullScreen: !isFullScreen });
    if (attachments) {
      this.setState({ attachments });
    }
  };

  private onCompleteDetails = (assetId: number): void => {
    const { navigation, setAssetId, setEditPropertyFlow } = this.props;
    setAssetId(assetId);
    setEditPropertyFlow(true);
    // @ts-ignore
    navigation.navigate(ScreensKeys.PropertyPostStack, {
      screen: ScreensKeys.AddProperty,
      params: { previousScreen: ScreensKeys.Dashboard },
    });
  };

  private onCloseDeleteView = (): void => {
    this.setState({
      isDeleteProperty: false,
    });
  };

  private onDeleteProperty = async (id: number): Promise<void> => {
    const { t, navigation } = this.props;
    this.setState({ isLoading: true });
    try {
      await AssetRepository.deleteAsset(id);
      this.setState(
        {
          isDeleteProperty: false,
          isLoading: false,
        },
        () => {
          AlertHelper.success({ message: t('property:propertyDeleted') });
          navigation.goBack();
        }
      );
    } catch (e) {
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
      this.setState({ isDeleteProperty: false, isLoading: false });
    }
  };

  private onSelectMenuItem = (value: string): void => {
    const {
      navigation,
      setSelectedPlan,
      setAssetId,
      getAssetById,
      setEditPropertyFlow,
      toggleEditPropertyFlowBottomSheet,
    } = this.props;
    const {
      propertyData: {
        id,
        lastVisitedStep: {
          listing: { type },
        },
        assetStatusInfo,
        assetGroup,
      },
      propertyData,
    } = this.state;
    let params;
    setSelectedPlan({ id, selectedPlan: type });
    setAssetId(id);
    getAssetById();

    switch (value) {
      case MenuItems.EDIT_LISTING:
        // @ts-ignore
        navigation.navigate(ScreensKeys.PropertyPostStack, {
          screen: ScreensKeys.AssetListing,
          params: { previousScreen: ScreensKeys.Dashboard, isEditFlow: true },
        });
        break;
      case MenuItems.EDIT_PROPERTY:
        setEditPropertyFlow(true);
        if (assetStatusInfo && assetStatusInfo.status) {
          params = {
            status: assetStatusInfo.status,
          };
          if (assetStatusInfo.status === 'APPROVED') {
            toggleEditPropertyFlowBottomSheet(true);
          }
        }
        // @ts-ignore
        navigation.navigate(ScreensKeys.PropertyPostStack, {
          screen: ScreensKeys.PostAssetDetails,
          ...(params && { params }),
        });
        break;
      case MenuItems.DELETE_PROPERTY:
        this.setState({
          isDeleteProperty: true,
        });
        break;
      case MenuItems.MANAGE_TENANT:
        navigation.navigate(ScreensKeys.ManageTenantScreen, {
          assetDetail: propertyData,
        });
        break;
      case MenuItems.EDIT_LEASE:
        if (assetStatusInfo?.leaseTransaction) {
          navigation.navigate(ScreensKeys.UpdateLeaseScreen, {
            transactionId: assetStatusInfo.leaseTransaction.id,
            assetGroup: assetGroup.code,
            user: assetStatusInfo.leaseTenantInfo.user,
          });
        }
        break;
      default:
        FunctionUtils.noop();
    }
  };

  private handleTabAction = (key: Tabs): void => {
    const { navigation } = this.props;
    const { propertyData, isFromTenancies } = this.state;
    if (isEmpty(propertyData)) {
      return;
    }
    const { formattedProjectName, id, assetStatusInfo } = propertyData;

    const param = {
      isFromPortfolio: true,
      isFromTenancies: isFromTenancies ?? false,
      screenTitle: formattedProjectName,
      propertyId: id,
    };
    switch (key) {
      case Tabs.TICKETS:
        navigation.navigate(ScreensKeys.ServiceTicketScreen, param);
        break;
      case Tabs.OFFERS:
        navigation.navigate(ScreensKeys.OfferDetail, param);
        break;
      case Tabs.REVIEWS:
        navigation.navigate(ScreensKeys.AssetReviewScreen, {
          ...param,
          leaseListingId: assetStatusInfo?.leaseListingId ?? null,
          saleListingId: assetStatusInfo?.saleListingId ?? null,
        });
        break;
      case Tabs.SITE_VISITS:
        navigation.navigate(ScreensKeys.PropertyVisits, param);
        break;
      case Tabs.FINANCIALS:
        navigation.navigate(ScreensKeys.AssetFinancialScreen, param);
        break;
      case Tabs.MESSAGES:
        navigation.navigate(ScreensKeys.ChatScreen, param);
        break;
      case Tabs.DOCUMENTS:
        navigation.navigate(ScreensKeys.DocumentScreen, param);
        break;
      case Tabs.TENANT_HISTORY:
        navigation.navigate(ScreensKeys.TenantHistoryScreen, param);
        break;
      case Tabs.DETAILS:
        navigation.navigate(ScreensKeys.AssetDetailScreen, { ...param, property: propertyData });
        break;
      default:
        navigation.navigate(ScreensKeys.ComingSoonScreen, { title: key, tabHeader: formattedProjectName });
    }
  };

  private getAssetDetail = async (): Promise<void> => {
    const {
      assetPayload: { asset_id, assetType, listing_id },
      setCurrentChatDetail,
      setCurrentOfferPayload,
      getAsset,
      setFilter,
    } = this.props;

    if (!asset_id) {
      return;
    }

    const payload: IPropertyDetailPayload = {
      asset_id,
      id: listing_id,
      type: assetType,
    };
    this.setState({ isLoading: true });
    try {
      const response = await PortfolioRepository.getPropertyDetail(payload);
      const info = response.assetStatusInfo;
      this.clearStates();
      this.setState(
        {
          propertyData: response,
          isLoading: false,
          isFromTenancies: !response.isAssetOwner,
        },
        () => {
          if (info && info.leaseTransaction && info.leaseTransaction.id > 0) {
            setCurrentChatDetail({
              groupName: response.projectName,
              groupId: info.leaseTransaction.messageGroupId,
            });
          }

          if (info && (info.leaseListingId || info.saleListingId)) {
            setCurrentOfferPayload({
              type: info.leaseListingId ? ListingType.LEASE_LISTING : ListingType.SALE_LISTING,
              listingId: info.leaseListingId || info.saleListingId || 0,
            });
            setFilter({ asset_transaction_type: info.leaseListingId && info.leaseListingId > 0 ? 0 : 1 });
            getAsset({ propertyTermId: info.leaseListingId || info.saleListingId || 0 });
          }
        }
      );
    } catch (e) {
      this.setState({ isLoading: false });
      AlertHelper.error({ message: ErrorUtils.getErrorMessage(e.details) });
    }
  };

  private getMenuList = (isListingCreated: boolean, isOccupied: boolean): IMenu[] => {
    const {
      t,
      route: { params },
    } = this.props;
    const { isFromTenancies } = this.state;
    let list;
    if (isFromTenancies) {
      return [];
    }

    if (params && isOccupied) {
      list = [
        { label: t('property:editLeaseTerm'), value: MenuItems.EDIT_LEASE },
        { label: t('property:manageTenants'), value: MenuItems.MANAGE_TENANT },
      ];
      return list;
    }

    list = [
      { label: t('property:editProperty'), value: MenuItems.EDIT_PROPERTY },
      { label: t('property:deleteProperty'), value: MenuItems.DELETE_PROPERTY },
    ];

    if (isListingCreated) {
      list.splice(1, 0, { label: t('property:editListing'), value: MenuItems.EDIT_LISTING });
    }

    return list;
  };

  private handleAction = (asset: Asset, payload: IClosureReasonPayload, param?: IListingParam): void => {
    const { navigation, setAssetId } = this.props;
    const { CancelListing, TerminateListing } = UpdatePropertyFormTypes;
    const { LEASE_TRANSACTION_TERMINATION } = ClosureReasonType;
    const formType = payload.type === LEASE_TRANSACTION_TERMINATION ? TerminateListing : CancelListing;
    const { id, isSubleased, assetStatusInfo } = asset;
    const startDate = DateUtils.getUtcFormatted(assetStatusInfo?.leaseTransaction.leaseEndDate ?? '', 'DD/MM/YYYY');

    setAssetId(id);

    if (param && param.hasTakeAction) {
      navigation.navigate(ScreensKeys.AssetPlanSelection, {
        isFromPortfolio: true,
        isSubleased,
        leaseUnit: assetStatusInfo?.leaseUnitId ?? undefined,
        startDate: DateUtils.getFutureDateByUnit(startDate, 1, 'days'),
      });
    } else {
      navigation.navigate(ScreensKeys.UpdatePropertyScreen, { formType, payload, param, assetDetail: asset });
    }
  };

  private updateSlide = (slideNumber: number): void => {
    this.setState({ activeSlide: slideNumber });
  };

  private handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
    this.clearStates();
  };

  private clearStates = (): void => {
    const { clearChatDetail, clearMessages, clearState } = this.props;
    clearChatDetail();
    clearMessages();
    clearState();
  };
}

const mapStateToProps = (state: IState): IStateProps => {
  return {
    assetPayload: PortfolioSelectors.getCurrentAssetPayload(state),
    asset: AssetSelectors.getAsset(state),
  };
};

const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const {
    setAssetId,
    setSelectedPlan,
    getAssetById,
    setEditPropertyFlow,
    toggleEditPropertyFlowBottomSheet,
  } = RecordAssetActions;
  const { clearAsset, getAsset } = AssetActions;
  const { clearChatDetail, clearMessages, setCurrentChatDetail } = CommonActions;
  const { setCurrentOfferPayload, setCompareDetail, clearState } = OfferActions;
  const { setFilter } = SearchActions;
  return bindActionCreators(
    {
      setAssetId,
      setSelectedPlan,
      getAssetById,
      setEditPropertyFlow,
      toggleEditPropertyFlowBottomSheet,
      clearAsset,
      clearChatDetail,
      clearMessages,
      setCurrentChatDetail,
      setCurrentOfferPayload,
      getAsset,
      setFilter,
      setCompareDetail,
      clearState,
    },
    dispatch
  );
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation(LocaleConstants.namespacesKey.assetPortfolio)(PropertyDetailScreen));

const styles = StyleSheet.create({
  flexOne: {
    flex: 1,
  },
  card: {
    borderRadius: 0,
  },
  label: {
    textAlign: 'center',
    color: theme.colors.darkTint3,
  },
  background: {
    backgroundColor: theme.colors.white,
  },
  tabView: {
    backgroundColor: theme.colors.white,
    paddingVertical: 20,
  },
  tabItem: {
    marginHorizontal: 16,
    alignItems: 'center',
  },
});
