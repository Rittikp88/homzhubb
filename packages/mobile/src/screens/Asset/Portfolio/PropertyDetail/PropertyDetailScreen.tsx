import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { connect } from 'react-redux';
import { TabBar, TabView } from 'react-native-tab-view';
import { PortfolioNavigatorParamList } from '@homzhub/mobile/src/navigation/BottomTabs';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import Icon, { icons } from '@homzhub/common/src/assets/icon';
import { theme } from '@homzhub/common/src/styles/theme';
import { Text } from '@homzhub/common/src/components';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { PortfolioSelectors } from '@homzhub/common/src/modules/portfolio/selectors';
import { AnimatedProfileHeader, FullScreenAssetDetailsCarousel, Loader } from '@homzhub/mobile/src/components';
import AssetCard from '@homzhub/mobile/src/components/organisms/AssetCard';
import SiteVisitTab from '@homzhub/mobile/src/components/organisms/SiteVisitTab';
import NotificationTab from '@homzhub/mobile/src/screens/Asset/Portfolio/PropertyDetail/NotificationTab';
import DummyView from '@homzhub/mobile/src/screens/Asset/Portfolio/PropertyDetail/DummyView';
import Documents from '@homzhub/mobile/src/screens/Asset/Portfolio/PropertyDetail/Documents';
import TenantHistoryScreen from '@homzhub/mobile/src/screens/Asset/Portfolio/PropertyDetail/TenantHistoryScreen';
import { Asset } from '@homzhub/common/src/domain/models/Asset';
import { Attachment } from '@homzhub/common/src/domain/models/Attachment';
import { Tabs, Routes } from '@homzhub/common/src/constants/Tabs';

const TAB_LAYOUT = {
  width: theme.viewport.width - theme.layout.screenPadding * 2,
  height: theme.viewport.height,
};

interface IStateProps {
  propertyData: Asset | null;
}

interface IDetailState {
  attachments: Attachment[];
  isFullScreen: boolean;
  activeSlide: number;
  currentIndex: number;
}

interface IRoutes {
  key: string;
  title: string;
  icon: string;
}

type libraryProps = NavigationScreenProps<PortfolioNavigatorParamList, ScreensKeys.PropertyDetailScreen>;
type Props = WithTranslation & libraryProps & IStateProps;

export class PropertyDetailScreen extends Component<Props, IDetailState> {
  public state = {
    isFullScreen: false,
    activeSlide: 0,
    attachments: [],
    currentIndex: 0,
  };

  public render = (): React.ReactNode => {
    const { t, propertyData } = this.props;
    const { currentIndex } = this.state;
    if (!propertyData) {
      return null;
    }
    return (
      <>
        <AnimatedProfileHeader title={t('portfolio')}>
          <>
            <View style={styles.header}>
              <Icon
                name={icons.leftArrow}
                size={20}
                color={theme.colors.primaryColor}
                onPress={this.handleIconPress}
                testID="icnBack"
              />
              <Text type="small" textType="semiBold" style={styles.headerTitle}>
                {t('propertyDetails')}
              </Text>
            </View>
            <AssetCard assetData={propertyData} isDetailView enterFullScreen={this.onFullScreenToggle} />
            <TabView
              lazy
              renderLazyPlaceholder={(): React.ReactElement => <Loader visible />}
              removeClippedSubviews
              swipeEnabled={false}
              initialLayout={TAB_LAYOUT}
              renderScene={({ route }): React.ReactElement => this.renderTabScene(route)}
              onIndexChange={this.handleIndexChange}
              renderTabBar={(props): React.ReactElement => {
                const {
                  navigationState: { index, routes },
                } = props;
                const currentRoute = routes[index];
                return (
                  <TabBar
                    {...props}
                    scrollEnabled
                    style={{ backgroundColor: theme.colors.white }}
                    indicatorStyle={{ backgroundColor: theme.colors.blue }}
                    renderIcon={({ route }): React.ReactElement => {
                      const isSelected = currentRoute.key === route.key;
                      return (
                        <Icon
                          name={route.icon}
                          color={isSelected ? theme.colors.blue : theme.colors.darkTint3}
                          size={22}
                        />
                      );
                    }}
                    renderLabel={({ route }): React.ReactElement => {
                      const isSelected = currentRoute.key === route.key;
                      return (
                        <Text type="small" style={{ color: isSelected ? theme.colors.blue : theme.colors.darkTint3 }}>
                          {route.title}
                        </Text>
                      );
                    }}
                  />
                );
              }}
              navigationState={{
                index: currentIndex,
                routes: Routes,
              }}
            />
          </>
        </AnimatedProfileHeader>
        {this.renderFullscreenCarousel()}
      </>
    );
  };

  private renderTabScene = (route: IRoutes): React.ReactElement => {
    const { navigation } = this.props;
    switch (route.key) {
      case Tabs.NOTIFICATIONS:
        // TODO: Figure-out something to resolve this error
        // @ts-ignore
        return <NotificationTab />;
      case Tabs.SITE_VISITS:
        return (
          <View style={{ backgroundColor: theme.colors.white }}>
            <SiteVisitTab onReschedule={this.navigateToBookVisit} navigation={navigation} />
          </View>
        );
      case Tabs.DOCUMENTS:
        return <Documents />;
      case Tabs.TENANT_HISTORY:
        return <TenantHistoryScreen />;
      default:
        return <DummyView />;
    }
  };

  private renderFullscreenCarousel = (): React.ReactNode => {
    const { isFullScreen, activeSlide, attachments } = this.state;
    if (!isFullScreen) return null;
    return (
      <FullScreenAssetDetailsCarousel
        onFullScreenToggle={this.onFullScreenToggle}
        activeSlide={activeSlide}
        data={attachments}
        updateSlide={this.updateSlide}
      />
    );
  };

  private onFullScreenToggle = (attachments?: Attachment[]): void => {
    const { isFullScreen } = this.state;
    this.setState({ isFullScreen: !isFullScreen });
    if (attachments) {
      this.setState({ attachments });
    }
  };

  private navigateToBookVisit = (isNew?: boolean): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.SearchStack, {
      screen: ScreensKeys.BookVisit,
      params: { isReschedule: !isNew },
    });
  };

  private handleIndexChange = (index: number): void => {
    this.setState({ currentIndex: index });
  };

  private updateSlide = (slideNumber: number): void => {
    this.setState({ activeSlide: slideNumber });
  };

  private handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };
}

const mapStateToProps = (state: IState): IStateProps => {
  return {
    propertyData: PortfolioSelectors.getAssetById(state),
  };
};

export default connect(
  mapStateToProps,
  null
)(withTranslation(LocaleConstants.namespacesKey.assetPortfolio)(PropertyDetailScreen));

const styles = StyleSheet.create({
  header: {
    backgroundColor: theme.colors.white,
    flexDirection: 'row',
    alignItems: 'center',
    padding: 10,
    borderBottomRightRadius: 0,
    borderBottomLeftRadius: 0,
  },
  headerTitle: {
    color: theme.colors.darkTint1,
    marginLeft: 12,
  },
});
