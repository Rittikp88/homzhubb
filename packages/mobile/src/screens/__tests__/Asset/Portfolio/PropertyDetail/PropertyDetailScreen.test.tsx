import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PropertyDetailScreen } from '@homzhub/mobile/src/screens/Asset/Portfolio/PropertyDetail/PropertyDetailScreen';
import { TenanciesAssetData } from '@homzhub/common/src/mocks/AssetData';

describe('Property Detail Screen', () => {
  let component: ShallowWrapper;
  let props: any;

  beforeEach(() => {
    props = {
      navigation: {
        goBack: jest.fn(),
      },
      route: {
        params: {
          propertyData: TenanciesAssetData[0],
        },
      },
    };
    component = shallow(<PropertyDetailScreen {...props} t={(key: string): string => key} />);
  });

  it('should render snapshot', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should call goBack', () => {
    // @ts-ignore
    component.find('[testID="icnBack"]').prop('onPress')();
    expect(props.navigation.goBack).toBeCalled();
  });
});
