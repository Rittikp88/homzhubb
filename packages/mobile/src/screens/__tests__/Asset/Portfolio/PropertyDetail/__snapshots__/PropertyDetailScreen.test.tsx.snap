// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property Detail Screen should render snapshot 1`] = `
<AnimatedProfileHeader
  title="portfolio"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderBottomLeftRadius": 0,
        "borderBottomRightRadius": 0,
        "flexDirection": "row",
        "padding": 10,
      }
    }
  >
    <Icon
      allowFontScaling={false}
      color="#0084F8"
      name="left-arrow"
      onPress={[Function]}
      size={20}
      testID="icnBack"
    />
    <Text
      style={
        Object {
          "color": "#001B36",
          "marginLeft": 12,
        }
      }
      textType="semiBold"
      type="small"
    >
      propertyDetails
    </Text>
  </View>
  <withI18nextTranslation(AssetCard)
    assetData={
      Object {
        "assetStatusInfo": Object {
          "leaseTenantInfo": Object {},
          "leaseTransaction": Object {},
          "tag": Object {},
        },
        "asset_group": Object {
          "id": 1,
          "name": "Residential",
        },
        "asset_type": Object {
          "id": 5,
          "name": "Studio Apartment",
        },
        "attachments": Array [
          Object {
            "file_name": "prof.jpg",
            "is_cover_image": true,
            "link": "https://homzhub-bucket.s3.amazonaws.com/asset_images/8e8c48fc-c089-11ea-8247-34e12d38d70eprof.jpg",
            "media_attributes": Object {},
            "media_type": "IMAGE",
          },
        ],
        "block_number": "B",
        "carpet_area": 2500,
        "carpet_area_unit": Object {
          "id": 1,
          "label": "Square feet",
          "name": "SQ_FT",
          "title": "Sq.ft",
        },
        "construction_year": null,
        "description": "",
        "digital_id": "8fad8389-6893-411a-bd4a-4b1575523d4a",
        "floor_number": 2,
        "furnishing": "SEMI",
        "id": 5,
        "is_gated": null,
        "latitude": 25.6207722,
        "longitude": 85.1280892,
        "notifications": Object {
          "count": 5,
        },
        "progress_percentage": 74,
        "project_name": "House 2",
        "serviceTickets": Object {
          "count": 0,
        },
        "spaces": Array [
          Object {
            "count": 2,
            "id": 1,
            "name": "Bedroom",
          },
          Object {
            "count": 1,
            "id": 2,
            "name": "Bathroom",
          },
        ],
        "total_floors": 3,
        "unit_number": "103",
        "verification_documents": Array [],
        "verifications": Object {
          "description": "**What is this verification?**
Sed ut perspiciatis unde omnis iste natus error sit voluptatem
accusantium doloremque laudantium.

Nemo enim ipsam voluptatem quia voluptas sit aspernatur
aut odit aut fugit. Neque porro quisquam est,
qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit,
sed quia non numquam eius modi tempora incidunt ut labore et dolore
magnam aliquam quaerat voluptatem.",
          "verification_status": 2,
        },
      }
    }
    isDetailView={true}
  />
  <TopTabs />
</AnimatedProfileHeader>
`;
